subgraphs/enzyme-core/mappings/v4/ExternalPositionManager.ts

```diff
@@ -1809,6 +1824,163 @@ export function handleCallOnExternalPositionExecutedForFund(event: CallOnExterna
     return;
   }
 
+  if (type.label == 'PENDLE_V2') {
+    if (actionId == PendleV2ActionId.BuyPrincipalToken) {
+      let decoded = ethereum.decode(
+        '(address,address,uint256,tuple(uint256,uint256,uint256,uint256,uint256),uint256)',
+        event.params.actionArgs,
+      );
+
+      if (decoded == null) {
+        return;
+      }
+
+      let tuple = decoded.toTuple();
+
+      let market = tuple[0].toAddress();
+      let depositToken = ensureAsset(tuple[1].toAddress());
+      let depositAmount = toBigDecimal(tuple[2].toBigInt(), depositToken.decimals);
+
+      let pendleMarket = ensurePendleV2AllowedMarket(vault, market);
+
+      let position = usePendleV2Position(event.address.toHex());
+      position.principalTokenHoldings = arrayUnique(position.principalTokenHoldings.concat([pendleMarket.id]));
+      position.save();
+
+      let change = createPendleV2PositionChange(event.params.externalPosition, 'BuyPrincipalToken', vault, event);
+      change.assetAmount = createAssetAmount(depositToken, depositAmount, denominationAsset, 'pendle-buy-pt', event).id;
+      change.market = pendleMarket.id;
+      change.save();
+    }
+
+    if (actionId == PendleV2ActionId.SellPrincipalToken) {
+      let decoded = ethereum.decode('(address,address,uint256,uint256)', event.params.actionArgs);
+
+      if (decoded == null) {
+        return;
+      }
+
+      let tuple = decoded.toTuple();
+
+      let market = tuple[0].toAddress();
+      let withdrawalToken = ensureAsset(tuple[1].toAddress());
+      let withdrawalAmount = toBigDecimal(tuple[2].toBigInt(), withdrawalToken.decimals);
+
+      let pendleMarket = ensurePendleV2AllowedMarket(vault, market);
+
+      let balance = tokenBalance(Address.fromString(pendleMarket.principalToken), event.params.externalPosition);
+      if (balance && balance.isZero()) {
+        let position = usePendleV2Position(event.params.externalPosition.toHex());
+        //  TODO: check if this is correct. There could be several markets for a single principle token.
+        position.principalTokenHoldings = arrayDiff(position.principalTokenHoldings, [pendleMarket.id]);
```

Comment:
yeah, it looks like there can be a situation where a single PT is on multiple markets. So probably what we would like to do here is to filter out from `principalTokenHoldings` all of the markets that include this 0 principle token
<img width="1043" alt="Screenshot 2024-04-11 at 15 51 33" src="https://github.com/enzymefinance/subgraphs/assets/67273035/48fcbec6-64bb-416c-a43b-ad11e5362b68">